clang/test/SemaCXX/constexpr-math.cpp

```diff
@@ -0,0 +1,57 @@
+// RUN: %clang_cc1 -DWIN -verify -std=c++23 -fsyntax-only  %s
+// RUN: %clang_cc1 -verify -std=c++23 -fsyntax-only  %s
+
+// expected-no-diagnostics
+
+
+#ifdef WIN
+#define INFINITY ((float)(1e+300 * 1e+300))
+#define NAN      (-(float)(INFINITY * 0.0F))
+#else
+#define NAN (__builtin_nanf(""))
+#define INFINITY (__builtin_inff())
+#endif
+
+extern "C" void abort() noexcept;
+extern "C" int write(int, const void*, unsigned long);
+
+#define assert(condition) \
+  do {			    \
+    if (!(condition)) {					\
+      write(2, "Assertion failed: ", 18);		\
+      write(2, #condition, sizeof(#condition) - 1);	\
+      write(2, "\n", 1);				\
+      abort();						\
+    }							\
+  } while (false)
+
+int main() {
+    int i;
+
+    // fmin
+    static_assert(__builtin_fmin(15.24, 1.3) == 1.3, "");
+    static_assert(__builtin_fmin(-0.0, +0.0) == -0.0, "");
+    static_assert(__builtin_fmin(+0.0, -0.0) == -0.0, "");
+    assert(__builtin_isnan(__builtin_fminf(NAN,NAN)));
+    assert(__builtin_isnan(__builtin_fminf(NAN, -1)));
+    assert(__builtin_isnan(__builtin_fminf(-INFINITY, 0)));
+    assert(__builtin_iszero(__builtin_fminf(+INFINITY, 0)));
+
+    // frexp
+    static_assert(__builtin_frexp(123.45, &i) == 0.96445312500000002);
```

Comment:
This should not be a constant expression. This line is not supposed to compile.
`i` is supposed to be written to, but it lives outside of the constant expression evaluation.

Reply:
I am reading from `https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p0533r9.pdf` that the return type of __builtin_frexp() can be constexpr. 
This is the result of the code I have added in VisitCallExpr (`case Builtin::BI__builtin_frexpf:
case Builtin::BI__builtin_frexp: {}  that probably needs to be guarded by getLangOpts().CPlusPlus23).

Reply:
Just because a function is `constexpr` does not mean an invocation of it would be allowed in a constant expression even if the arguments were constant expressions.

The specific issue is that the semantics of `frexp` has it write a value through it's second parameter. In order to modify the stored value of an object in constant evaluation, the object has to begin its lifetime in that constant evaluation.

Reply:
It seems that we should write something like
```C++
struct { double num_; int expo_; } result;
result.num_ = frexp(x, &result.expo_);
return result;
```

Reply:
Still working on this. Can you please review the implementation of frexp I have added in `ExprConstant.cpp`?
@frederick-vs-ja  are you suggesting using something like this for the use of `frexp` in this test?
```
struct FrexpResult {
  double fraction;
  int exponent;
};

FrexpResult simulateFrexp(double value) {
  FrexpResult result;
  result.fraction = __builtin_frexp(value, &result.exponent);
  return result;
}

int main () {
  double value = 123.456;
  FrexpResult result = simulateFrexp(123.456);
  
  return 0;
}

```
Still no way of asserting what the return values of fraction and exponent are from calls to `simulateFrex`p with various arguments. Any suggestions?

Reply:
@zahiraam, make `simulateFrexp` a `constexpr` function (maybe with `float` and `long double` overloads as well).

```
template <typename T>
struct FrexpResult {
  constexpr FrexpResult(T fraction, int exponent) : fraction(fraction), exponent(exponent) {}
  T fraction;
  int exponent;
  friend constexpr bool operator<=>(const FrexpResult &, const FrexpResult &) = default;
};

constexpr auto q() { return FrexpResult{0.5, 0}; };
static_assert(q() == FrexpResult{0.5, 0});
```

Reply:
Here is an example of checking both the fraction and exponent values by "identity" instead of by comparisons.

```cpp
template <auto, auto> constexpr bool is_same_val = false;
template <auto X> constexpr bool is_same_val<X, X> = true;

template <typename T>
struct FrexpResult {
  T fraction;
  int exponent;
};

template <typename T>
constexpr auto simulateFrexp(T value) {
  FrexpResult<T> result;
  result.fraction = __builtin_frexp(value, &result.exponent);
  return result;
}

static_assert(is_same_val<simulateFrexp(1.), FrexpResult{0.5, 1}>);
```

https://godbolt.org/z/Whx6o5Y7E